File: elixir/triangle/triangle.exs
defmodule Triangle do
  @type kind :: :equilateral | :isosceles | :scalene

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: { :ok, kind } | { :error, String.t }
  def kind(a, b, c) do
    cond do
      a <= 0 -> { :error, "all side lengths must be positive" }
      b <= 0 -> { :error, "all side lengths must be positive" }
      c <= 0 -> { :error, "all side lengths must be positive" }
      a + b <= c -> {:error, "side lengths violate triangle inequality"}
      a + c <= b -> {:error, "side lengths violate triangle inequality"}
      b + c <= a -> {:error, "side lengths violate triangle inequality"}
      true -> do_kind(a,b,c)
    end
  end

  defp do_kind(a, a, a), do: {:ok, :equilateral}
  defp do_kind(a, a, _), do: {:ok, :isosceles}
  defp do_kind(_, a, a), do: {:ok, :isosceles}
  defp do_kind(a, _, a), do: {:ok, :isosceles}
  defp do_kind(_, _, _), do: {:ok, :scalene}
end

File: go/triangle/triangle.go
// Determine if a triangle is equilateral, isosceles, or scalene.
package triangle

import "math"

type Kind int

const (
	NaT = iota // not a triangle
	Equ        // equilateral
	Iso        // isosceles
	Sca        // scalene
)

// Determine the kind of trangle based on the provided side lengths
func KindFromSides(a, b, c float64) Kind {
	if isInvalid(a, b, c) || inequality(a, b, c) {
		return NaT
	}

	if a == b && b == c {
		return Equ
	}

	if (a == b) || (b == c) || (a == c) {
		return Iso
	}

	return Sca
}

func isInvalid(nums ...float64) bool {
	for _, num := range nums {
		if math.IsNaN(num) || math.IsInf(num, 0) || num <= 0 {
			return true
		}
	}
	return false
}

func inequality(a, b, c float64) bool {
	return a+b < c || a+c < b || b+c < a
}
